pkg/retriever/cidretrieval.go

```diff
@@ -158,12 +157,12 @@ func (ret *retrieval) collectResults(candidates []RetrievalCandidate) (*Retrieva
 		}
 		// have we got all responses but no success?
 		finishedCount++
-		if finishedCount >= len(candidates) {
+		if finishedCount >= cap(resultChan) {
```

Comment:
I think there's a race in using `cap` because while we're inside this loop we may have another entry queued up.

Thinking this through with an example, I believe this is a possibility: if we have a capacity of 4 and have already `finishedCount` of 2 by this point, under normal circumstances we'd expect `cap(resultChan)` to yield 4 so we're fine, but what if we had a freak condition where 2 additional results got pushed into the channel just before getting to this point, so I believe our `cap(resultChan)` would go down to 2 and our condition would pass and we'd end up ignoring the two things left in `resultChan`?

Reply:
`cap(resultChan)` will return the total capacity of a buffered channel. In this case, the buffer size of the `resultChan` channel is [the number of candidates returned from `findCandidates()`](https://github.com/filecoin-project/lassie/pull/13/files/f688a8ec5ad4717fc5216275007ba4c24082fa78#diff-963da40016fcc9e7b86e9c736013c866539c7235fd6ee299189f5f8048c142d3R135).

To continue using your example, if we send a value to the `resultChan` channel before this point, `cap(resultChan)` will still return 4 for a buffered channel. Capacity for a buffered channel is just the total buffer size, not the remaining amount of space left in the channel. We make a buffered channel of size 4, `cap()` will always return 4. Here is a [go playground](https://go.dev/play/p/nRgfXJbJC1L) showing this behavior.